File: src/main/java/ui/MainView.java
package ui;

import com.nospawnn.operations.Operation;
import com.nospawnn.operations.OperationsList;

import javax.swing.*;
import javax.swing.event.DocumentEvent;
import javax.swing.event.DocumentListener;
import javax.swing.tree.DefaultMutableTreeNode;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.util.ArrayList;
import java.util.List;

public class MainView extends JFrame {
    private JPanel contentPane;
    private JTree sidePanelTree;
    private JTextArea inputTextArea;
    private JTextArea outputTextArea;
    private JTextField searchTextField;
    private JList<String> displayedStepsList;
    private JScrollPane stepsListScrollPane;
    private JPanel textPane;
    private JScrollPane operationsScrollPane;
    private JPanel operationsPane;
    private JPanel stepsPane;

    private final List<Operation> steps;

    public MainView() {
        steps = new ArrayList<>();

        setTitle("CyberDev");
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setContentPane(contentPane);

        pack();
        setLocationRelativeTo(null);
    }

    private void createUIComponents() {
        createStepsList();
        createRecipeTree();
    }

    private void createStepsList() {
        DefaultListModel<String> stepsListModel = new DefaultListModel<>();
        displayedStepsList = new JList<>(stepsListModel);
    }

    private void createRecipeTree() {
        var treeEntries = OperationsList.INSTANCE.getOperations();

        var root = new DefaultMutableTreeNode(null);
        for (var entry : treeEntries.entrySet()) {
            var folderNode = new DefaultMutableTreeNode(entry.getKey());
            for (var op : entry.getValue()) folderNode.add(new DefaultMutableTreeNode(op.getListText()));
            root.add(folderNode);
        }

        sidePanelTree = new JTree(root);
        sidePanelTree.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                if (e.getClickCount() != 2) return;

                var selected = (DefaultMutableTreeNode) sidePanelTree.getLastSelectedPathComponent();

                if (selected == null || !selected.isLeaf()) return;

                var optOp = treeEntries.get(selected.getParent().toString())
                        .stream()
                        .filter(o -> o.getListText().equals(selected.toString()))
                        .findFirst();

                if (optOp.isEmpty()) return;

                var op = optOp.get();
                ((DefaultListModel<String>) displayedStepsList.getModel()).addElement(op.getListText());
                steps.add(op);

                updateOutput();
            }
        });

        displayedStepsList.addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                if (e.getClickCount() != 2) return;

                var idxToRemove = displayedStepsList.getSelectedIndex();
                ((DefaultListModel<String>) displayedStepsList.getModel()).remove(idxToRemove);
                steps.remove(idxToRemove);

                updateOutput();
            }
        });

        inputTextArea = new JTextArea();
        inputTextArea.getDocument().addDocumentListener(new DocumentListener() {
            @Override
            public void insertUpdate(DocumentEvent e) {
                updateOutput();
            }

            @Override
            public void removeUpdate(DocumentEvent e) {
                updateOutput();
            }

            @Override
            public void changedUpdate(DocumentEvent e) {
            }
        });
    }

    private void updateOutput() {
        var text = inputTextArea.getText();

        for (var step : steps)
            text = step.perform(text);

        outputTextArea.setText(text);
    }

}

File: src/main/kotlin/ui/MainView.kt
package com.nospawnn.ui

import com.nospawnn.operations.Operation
import java.awt.Dimension
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import java.awt.Insets
import javax.swing.DefaultListModel
import javax.swing.JFrame
import javax.swing.JList
import javax.swing.JPanel

class MainView : JFrame() {
    private val textPane = TextPane()
    private val stepsList = JList<Operation>(DefaultListModel<Operation>())
    private val argumentsEditorPane = ArgumentEditorPane()
    private val stepsPane = StepsListPane(stepsList, argumentsEditorPane)
    private val operationsPane = OperationsPane(stepsList, argumentsEditorPane)

    private var inputText = ""

    init {
        createUI()
    }

    private fun createUI() {
        this.apply {
            title = "CyberDev"
            defaultCloseOperation = EXIT_ON_CLOSE
            minimumSize = Dimension(1280, 720)
            contentPane = JPanel().apply {
                layout = GridBagLayout()
            }
        }

        val gbc = GridBagConstraints().apply {
            fill = GridBagConstraints.BOTH
            gridx = GridBagConstraints.RELATIVE
        }

        gbc.gridy = 0
        gbc.weightx = 0.25
        gbc.gridwidth = 3
        gbc.weighty = 1.0
        gbc.insets = Insets(10, 10, 10, 10)
        contentPane.add(operationsPane, gbc)

        gbc.weightx = 0.1
        gbc.gridwidth = 1
        gbc.insets = Insets(0, 0, 0, 0)
        val stepsAndArgsPane = JPanel().apply {
            layout = GridBagLayout()
            add(stepsPane, gbc)
            gbc.gridy = 1
            add(argumentsEditorPane, gbc)
        }
        gbc.gridy = 0
        gbc.insets = Insets(10, 10, 10, 10)
        contentPane.add(stepsAndArgsPane, gbc)

        gbc.weighty = 1.0
        gbc.weightx = 0.65
        gbc.gridwidth = 3
        contentPane.add(textPane, gbc)

        pack()
        setLocationRelativeTo(null)
    }

    fun updateOutput() {
        inputText = textPane.getInput()

        var output = inputText
        for (op in (stepsList.model as DefaultListModel<Operation>).elements()) {
            try {
                output = op.perform(output, argumentsEditorPane.getArgsFor(op.toString()))
            } catch (e: Exception) {
                println(e)
                textPane.displayOutputText("Invalid input for step \"${op}\"")
                return
            }
        }

        textPane.displayOutputText(output)
    }
}